crates/auto-hash-map/src/map.rs

```diff
@@ -0,0 +1,585 @@
+use std::{
+    borrow::Borrow,
+    collections::{hash_map::DefaultHasher, HashMap},
+    fmt::{Debug, Formatter},
+    hash::{BuildHasher, BuildHasherDefault, Hash},
+    marker::PhantomData,
+};
+
+use serde::{
+    de::{MapAccess, Visitor},
+    ser::SerializeMap,
+    Deserialize, Deserializer, Serialize, Serializer,
+};
+
+use crate::{MAX_LIST_SIZE, MIN_HASH_SIZE};
+
+#[derive(Clone)]
+pub enum AutoMap<K, V, H = BuildHasherDefault<DefaultHasher>> {
+    List(Vec<(K, V)>),
+    Map(Box<HashMap<K, V, H>>),
+}
+
+impl<K, V, H> Default for AutoMap<K, V, H> {
+    fn default() -> Self {
+        Self::List(Default::default())
+    }
+}
+
+impl<K: Debug, V: Debug, H> Debug for AutoMap<K, V, H> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_map().entries(self.iter()).finish()
+    }
+}
+
+impl<K, V> AutoMap<K, V, BuildHasherDefault<DefaultHasher>> {
```

Comment:
It can't be in `impl<K, V, H: BuildHasher> AutoMap<K, V, H> {` since that would leave `H` not inferable. `new()` always uses the RandomState.

That's equal to https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.new